### PR TITLE
Add method to list objects

### PR DESCRIPTION
**Pull request by malthe**
See https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjectsV2.html for reference.

This fixes issue #8.

### CODE REVIEW

**Review by dizballanze — APPROVED** [review]

**Comment by mosquito** [comment]
Excellent work. Please fix the linter issues.

**Comment by dizballanze** [comment]
Thanks! Add an example to the README, please.

**Comment by malthe** [comment]
> Excellent work. Please fix the linter issues.

There's one linting issue remaining, but it's in existing code and I didn't quite know how to fix it.

**Review by mosquito — CHANGES_REQUESTED** [review]
Waiting for linter fixes

File: aiohttp_s3_client/xml.py [hunk 1]
```diff
@@ -30,3 +41,37 @@ def create_complete_upload_request(parts: List[Tuple[int, str]]) -> bytes:
...
+            elif tag == "LastModified":
+                assert text[-1] == "Z"
+                last_modified = datetime.fromisoformat(text[:-1]).replace(
+                    tzinfo=timezone.utc
+                )
+            elif tag == "Size":
+                size = int(text)
+            elif tag == "StorageClass":
+                storage_class = intern(text)
+        if etag and key and last_modified and size is not None and \
```
Comment (mosquito) [inline aiohttp_s3_client/xml.py]: Hint: Add parentheses and reformat this as a block

Comment (malthe) [reply aiohttp_s3_client/xml.py]: What do you use to autoformat this code? `pylava` wants me to do:
```python
        if (etag and key and last_modified and size is not None and
           storage_class):
```
Three spaces? Looks weird to me.

Comment (mosquito) [reply aiohttp_s3_client/xml.py]: I mean something like that:

```python
if (
    etag and
    key and
    last_modified and
    size is not None and
    storage_class
):
```

IMHO this way makes visual split between conditions and looks more readable.

Comment (malthe) [reply aiohttp_s3_client/xml.py]: Agreed and done.

File: aiohttp_s3_client/xml.py [hunk 2]
```diff
@@ -1,10 +1,21 @@
-from typing import List, Tuple
+from datetime import datetime
```
Comment (alvassin) [inline aiohttp_s3_client/xml.py]: `from datetime import datetime, timezone`

File: aiohttp_s3_client/client.py [hunk 3]
```diff
@@ -689,3 +692,64 @@ async def get_file_parallel(
...
+            object_name = f"/{bucket}"
+
+        while True:
+            async with self.get(str(object_name), params=params) as resp:
+                if resp.status != HTTPStatus.OK:
+                    raise AwsDownloadError(
+                        f"Got response for GET request for {object_name} "
+                        f"with prefix '{prefix}' "
+                        f"of a wrong status {resp.status}",
+                    )
```
Comment (alvassin) [inline aiohttp_s3_client/client.py:750]: Preposition `of` sounds strange for me, perhaps it's better to change it to `with`?

```
Got response with wrong status for GET request for {object_name} with prefix '{prefix}'?
```

**Comment by mosquito** [comment]
Please rebase this PR against master after #11

File: aiohttp_s3_client/client.py [hunk 4]
```diff
@@ -689,3 +692,64 @@ async def get_file_parallel(
...
+            limits the response to keys that begin with the specified
+            prefix
+        delimiter: a delimiter is a character you use to group keys
+        max_keys: maximum number of keys returned in the response
+        start_after: keys to start listing after
+        """
+
+        params = {
+            "list-type": "2",
+            "prefix": str(prefix),
```
Comment (alvassin) [inline aiohttp_s3_client/client.py]: URL would contain `?prefix=None` param by default (None as string). I think it is not supported behaviour when prefix argument is omitted.

**Review by alvassin — CHANGES_REQUESTED** [review]

File: aiohttp_s3_client/client.py [hunk 5]
```diff
@@ -691,3 +694,65 @@ async def get_file_parallel(
...
+        delimiter: a delimiter is a character you use to group keys
+        max_keys: maximum number of keys returned in the response
+        start_after: keys to start listing after
+        """
+
+        params = {
+            "list-type": "2",
+        }
+
+        if prefix:
```
Comment (mosquito) [inline aiohttp_s3_client/client.py:729]: Just as recommendation:

```python
params = dict(
    prefix=str(prefix) if prefix else None,
    delimiter=delimiter or None,
    max_keys=str(max_keys) if max_keys else None,
    start_after=start_after or None,
    object_name=f"/{bucket}",
)
params = {k: v for k, v in params.items() if v is not None}
```

Comment (malthe) [reply aiohttp_s3_client/client.py:729]: To be honest I'm not sure it is an improvement as such; one could also use a list comprehension to create the `params` dict.

I think the code as it is now is acceptable.

**Review by mosquito — APPROVED** [review]

**Review by dizballanze — APPROVED** [review]

**Review by alvassin — APPROVED** [review]

**Comment by mosquito** [comment]
@malthe released as [0.7.0](https://pypi.org/project/aiohttp-s3-client/0.7.0/)

**Pull request merged**